src/openfermion/measurements/_rdm_mapping_functions.py

```diff
@@ -0,0 +1,200 @@
+#   limitations under the License.
+
+"""Mapping RDMs to other RDMs"""
+import numpy
+from itertools import product
+
+
+def krond(i, j):
+    return 1 if i == j else 0
+
+
+def tpdm_to_opdm(tpdm, particle_number):
+    """
+    Contract a 2-RDM to a 1-RDM
+
+    Args:
+        tpdm (numpy.ndarray): The 2-RDM as a 4-index tensor. Indices follow the
+            internal convention of tpdm[p, q, r, s] ==
+            :math:`a_{p}^{\dagger}a_{q}^{\dagger}a_{r}a_{s}`.
+        particle_number (float): number of particles in the system
+
+    Returns:
+        opdm (numpy.ndarray): The 1-RDM contracted from the tpdm.
+    """
+    return numpy.einsum('prrq', tpdm) / (particle_number - 1)
+
+
+def tpdm_to_tqdm(tpdm, opdm):
+    """
+    Map from the 2-RDM to the 2-hole-RDM
+
+    Args:
+        tpdm (numpy.ndarray): The 2-RDM as a 4-index tensor. Indices follow the
+            internal convention of tpdm[p, q, r, s] ==
+            :math:`a_{p}^{\dagger}a_{q}^{\dagger}a_{r}a_{s}`.
+        opdm (numpy.ndarray): The 1-RDM as a 2-index tensor. Indices follow the
+            internal convention of opdm[p, q] ==
+            :math:`a_{p}^{\dagger}a_{q}`.
+
+    Returns:
+        tqdm (numpy.ndarray): The 2-hole matrix.
+    """
+    ldim = opdm.shape[0]
+    tqdm = numpy.zeros_like(tpdm)
+    for p, q, r, s in product(range(ldim), repeat=4):
+        term1 = opdm[p, s] * krond(q, r) + opdm[q, r] * krond(p, s)
+        term2 = -1 * (opdm[q, s] * krond(p, r) + opdm[p, r] * krond(q, s))
+        term3 = krond(q, s) * krond(p, r) - krond(p, s) * krond(q, r)
+        tqdm[s, r, q, p] = tpdm[p, q, r, s] - term1 - term2 - term3
+
+    return tqdm
+
+
+def tqdm_to_tpdm(tqdm, opdm):
+    """
+    Map from the 2-hole-RDM to the 2-RDM
+
+    Args:
+        tqdm (numpy.ndarray): The 2-hole-RDM as a 4-index tensor. Indices
+            follow the internal convention of tqdm[p, q, r, s] ==
+            :math:`a_{p}a_{q}a_{r}^{\dagger}a_{s}^{\dagger}`.
+        opdm (numpy.ndarray): The 1-RDM as a 2-index tensor. Indices follow the
+            internal convention of opdm[p, q] ==
+            :math:`a_{p}^{\dagger}a_{q}`.
+
+    Returns:
+        tpdm (numpy.ndarray): The 2-RDM matrix.
+    """
+    ldim = opdm.shape[0]
+    tpdm = numpy.zeros_like(tqdm)
+    for p, q, r, s in product(range(ldim), repeat=4):
+        term1 = opdm[p, s] * krond(q, r) + opdm[q, r] * krond(p, s)
+        term2 = -1 * (opdm[q, s] * krond(p, r) + opdm[p, r] * krond(q, s))
+        term3 = krond(q, s) * krond(p, r) - krond(p, s) * krond(q, r)
+        tpdm[p, q, r, s] = tqdm[r, s, p, q] + term1 + term2 + term3
+
+    return tpdm
+
+
+def tqdm_to_oqdm(tqdm, hole_number):
+    """
+    Map from 2-hole-RDM to 1-hole-RDM
+
+    Args:
+        tqdm (numpy.ndarray): The 2-hole-RDM as a 4-index tensor. Indices
+            follow the internal convention of tqdm[p, q, r, s] ==
+            :math:`a_{p}a_{q}a_{r}^{\dagger}a_{s}^{\dagger}`.
+        hole_number (float):
```

Comment:
Say what `hole_number` is (yeah it's pretty obvious), also the float cast should be done for `(hole_number - 1)` to be safe